ci-bench/src/main.rs

```diff
@@ -0,0 +1,309 @@
+    write_all_plaintext_bounded,
+};
+
+mod bench_lib;
+mod cli;
+mod util;
+
+/// The size in bytes of the plaintext sent in the transfer benchmark
+const TRANSFER_PLAINTEXT_SIZE: usize = 1024 * 1024;
+
+/// The amount of times a resumed handshake should be executed during benchmarking.
+///
+/// Handshakes with session resumption execute a very small amount of instructions (less than 200_000
+/// for some parameters), so a small difference in instructions accounts for a high difference in
+/// percentage (making the benchmark more sensitive to noise, because differences as low as 500
+/// instructions already raise a flag). Running the handshake multiple times gives additional weight
+/// to the instructions involved in the handshake, and less weight to noisy one-time setup code.
+///
+/// More specifically, great part of the noise in resumed handshakes comes from the usage of
+/// [`rustls::client::ClientSessionMemoryCache`] and [`rustls::server::ServerSessionMemoryCache`],
+/// which rely on a randomized `HashMap` under the hood (you can check for yourself by that
+/// `HashMap` by a `FxHashMap`, which brings the noise down to acceptable levels in a single run).
+const RESUMED_HANDSHAKE_RUNS: usize = 3;
+
+fn main() -> anyhow::Result<()> {
+    let benchmarks = get_benchmarks()?;
+
+    let cli = Cli::parse();
+    match cli.command {
+        Command::RunAll => {
+            let executable = std::env::args().next().unwrap();
+            let results = bench_lib::run_all(executable, &benchmarks)?;
+
+            // Output results in CSV (note: not using a library here to avoid extra dependencies)
+            for (name, instr_count) in results {
+                println!("{name},{instr_count}");
+            }
+        }
+        Command::RunSingle { index, side } => {
+            // `u32::MAX` is used as a signal to do nothing and return. By "running" an empty
+            // benchmark we can measure the startup overhead.
+            if index == u32::MAX {
+                return Ok(());
+            }
+
+            let bench = benchmarks
+                .get(index as usize)
+                .ok_or(anyhow::anyhow!("Benchmark not found: {index}"))?;
+
+            let mut stdin = io::stdin().lock();
+            let mut stdout = io::stdout().lock();
+
+            let result = match side {
+                Side::Server => run_bench_for_server(bench, &mut stdin, &mut stdout),
+                Side::Client => run_bench_for_client(bench, &mut stdin, &mut stdout),
+            };
+
+            result
+                .with_context(|| format!("{} crashed for {} side", bench.name(), side.as_str()))?;
+        }
+        Command::Compare {
+            baseline_input,
+            candidate_input,
+        } => {
+            bench_lib::compare(baseline_input, candidate_input)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Returns all benchmarks
+fn get_benchmarks() -> anyhow::Result<Vec<Benchmark>> {
```

Comment:
Naming nit: per the API guidelines, I would avoid the `get_()` prefix here.

Reply:
Will do